opcua/server/address_space.py

```diff
@@ -491,43 +491,75 @@ def empty(self):
 
     def dump(self, path):
         """
-        dump address space as binary to file
+        dump address space as binary to file; note that server must be stopped for this method to work
         """
-        s = shelve.open(path, "n", protocol = pickle.HIGHEST_PROTOCOL)
+        # prepare nodes in address space for being serialized
+        for nodeid, v in self._nodes.items():
+            ndata = self._nodes[nodeid]
+
+            # if the node has a reference to a method call, remove it so the object can be serialized
+            if ndata.call is not None:
+                self._nodes[nodeid].call = None
+
+        with open(path, 'wb') as f:
+            pickle.dump(self._nodes, f, pickle.HIGHEST_PROTOCOL)
+
+    def load(self, path):
+        """
+        load address space from file, overwriting everything in current address space
+        """
+        with open(path, 'rb') as f:
+            self._nodes = pickle.load(f)
+
+    def make_aspace_shelf(self, path):
+        """
+        make a shelf for storing nodes from the standard address space; nodes are moved to a cache
+        by the LazyLoadingDict class when they are accessed
+        """
+        s = shelve.open(path, "n", protocol=pickle.HIGHEST_PROTOCOL)
         for nodeid in self._nodes.keys():
             s[nodeid.to_string()] = self._nodes[nodeid]
         s.close()
```

Comment:
Pickle in `dump` and shelve are completely different. A shelve is a primitive dict style database where every item is pickled individually. Those pickled items are accessed by key string. They are only unpickled when they are accessed.
https://docs.python.org/3.4/library/shelve.html

The new dict is the shelve database (shelve is designed to act like a dict, read the docs) stored on disk. They way @stanti made this code it will only ever have the standard address space.